feat(cpininput): add c pin input

Shyrro opened the pull request:
Signed-off-by: Shyrro <zsahmane@gmail.com>

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes
      /start features)

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

None

## What is the new behavior?

Pin Input available


## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

- Due to `zag-js`, some behavior is not testable in `jest`/`cypress`. This is due to the fact that `zag` prevents some default behavior of the input fields to handle some events, which does not get along well with testing tools. 
This is kind of fine though, because the behavior is already tested on Zag's side so we won't have to worry too much about it.

- On another note, this uses the version `0.1.14` of `@zag-js/pin-input` instead of the latest version. This is because the latest release at this date, is `esm` only. Version should be upgraded when zag supports `cjs` again.

codebender828 commented on packages/c-pin-input/src/use-pin-input.ts:
```diff
@@ -0,0 +1,43 @@
+
+export const usePinInputMachine = (
+  params: PinInputProps,
+  emit: CallableFunction
```
Interesting. :) I didn't know there's a `CallableFunction` type haha 😄

codebender828 commented on packages/c-pin-input/src/use-pin-input.ts:
```diff
@@ -0,0 +1,43 @@
+      mask: params.mask,
+      blurOnComplete: params.blurOnComplete,
+      onChange: (value) => {
+        emit("change", value)
```
I think we can add support two way binding for the pin-input component using `v-model`. Perhaps at this `onChange` callback we can also call
```suggestion
        emit("change", value)
        emit("update:modelValue", value)
```

codebender828 commented on packages/c-pin-input/src/c-pin-input.tsx:
```diff
@@ -0,0 +1,139 @@
+export const CPinInput = defineComponent({
+  name: "CPinInput",
+  props: CPinInputProps,
+  emits: ["change", "invalid", "complete"],
```
@Shyrro I think we can add `v-model` support by including the `modelValue` prop and emitting the `update:modelValue` event.

That way we could listen to the value of the `modelValue` prop and then update the pin input machine's value using the [`api.value.setValue(props.modelValue)`](https://zagjs.com/components/react/pin-input#methods-and-properties) method.

This would allow the final consumer to listen to the `v-model` event if they want to do real-time validation with their forms.

An example of this is in the `CCheckboxGroup` component. In this example I used a writable `computed` variable to achieve this effect.

**`v-model` event handling**
https://github.com/chakra-ui/chakra-ui-vue-next/blob/50cab7ddf68db6db45e7d6c607a74597dd1639cf/packages/c-checkbox/src/checkbox-group.tsx#L51-L58

**Emitted value processing** (In this case it should be simple for the pin-input. No fancy processing is required here)
https://github.com/chakra-ui/chakra-ui-vue-next/blob/50cab7ddf68db6db45e7d6c607a74597dd1639cf/packages/c-checkbox/src/use-checkbox-group.ts#L3-L19

**Final consumption**
https://github.com/chakra-ui/chakra-ui-vue-next/blob/50cab7ddf68db6db45e7d6c607a74597dd1639cf/packages/c-checkbox/examples/with-checkbox-group.vue#L1-L19

codebender828 commented on packages/c-pin-input/tests/c-pin-input.cy.tsx:
```diff
@@ -0,0 +1,79 @@
+  it("should write correctly", () => {
+    render()
+
+    // We need to invoke val instead of use the "type" function
+    // Because Zag prevents some natural behavior of the input to replace it.
+    // So, invoking "val" is a way to type the value.
+    // The downside of this approach is that it won't naturally trigger the focus on the next element
```
I see. Is there any way we can test the public API of the `CInput` without manually mutating the `input` value with `jest` instead of `cypress`?

The reason for that is that it allows us to focus on the public API. Mutating the `<c-pin-input />` should update the `<c-pin-input-field/>` kind of disannuls the purpose of the test. Maybe jest could handle this, and we can skip this in cypress.

Shyrro replied:
Hi @codebender828 

Thanks again for your review ! 
We can indeed test the `CInput`alone, whether it's in `jest` or `cypress`. 
However, when it comes to the `CPinInput`, it's the same in `jest` or `cypress`... I'll try one last thing in `jest` to confirm.

I think best we can do is test that the `CInput` is working great from our side and leave the rest to Zag i think...

codebender828 reviewed (changes requested):
Thank you @Shyrro! Great PR! Zag really does reduce the amount of code being written by a lot! :)

I gave some feedback about:
1. `v-model` implementation of `c-pin-input`
2. Perhaps we should transfer the `c-pin-input` tests to Jest instead of Cypress. Reasons explained in the review

codebender828 commented:
Fun example BTW :)
<img width="441" alt="image" src="https://user-images.githubusercontent.com/21237954/195462416-4fe84662-df98-4e01-b3dc-b9a199682413.png">

Shyrro commented:
Hi @codebender828 

I updated the component to add `v-model` support. 
And for the test, i removed the useless test. I tried to add it in Jest, but it's the same issue. Normal inputs behave correctly, but inputs with zag props do no behave as we would like under testing environnements. In the playground it's perfect. So i removed the test and tested what i could.

codebender828 commented:
Thank you @Shyrro !
